File: src/game/systems/gameover.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local State = require 'game.state'
local Player = require 'game.components.player'

local aspect = Aspect.new({Player})
local GameOver = System:new('gameover', aspect)

function GameOver:update(dt)
  for _, entity in pairs(self.entities) do
    local player = entity:as(Player)

    if (player.lives == 0) then
      game:setState(State.GAMEOVER)
    end
  end
end

return GameOver

File: client2/src/game/systems/GameOver.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { Aspect } from '../../ecs/Aspect'
import { Player } from '../components/Player'
import { State } from '../State'

export class GameOverSystem extends System {
  static _id = 'GameOver'
  _id = GameOverSystem._id

  static _flag = SystemFlag.GameOver
  _flag = GameOverSystem._flag

  static _aspect = new Aspect([Player])
  _aspect = GameOverSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const player = entity.as(Player)

      if (player && player.lives === 0) {
        // game.setState(State.GAMEOVER)
      }
    })
  }
}
